From 04e12734afd9e3767a222785dabc9ec0ea94e490 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:05:18 -0400
Subject: [powerpc] pseries: Implement arch_get_random_long() based on H_RANDOM

Message-id: <1410545655-205645-90-git-send-email-dzickus@redhat.com>
Patchwork-id: 94244
O-Subject: [RHEL7 PATCH 089/626] powerpc/pseries: Implement arch_get_random_long() based on H_RANDOM
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit a489043f462639988f86c1cf49475580e9dba965
Author: Michael Ellerman <michael@ellerman.id.au>
Date:   Fri Oct 11 14:07:59 2013 +1100

    powerpc/pseries: Implement arch_get_random_long() based on H_RANDOM

    Add support for the arch_get_random_long() hook based on the H_RANDOM
    hypervisor call. We trust the hypervisor to provide us with random data,
    ie. we don't whiten it in anyway.

    Signed-off-by: Michael Ellerman <michael@ellerman.id.au>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---
 create mode 100644 arch/powerpc/platforms/pseries/rng.c

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/platforms/pseries/Makefile b/arch/powerpc/platforms/pseries/Makefile
index 6c61ec5..fbccac9 100644
--- a/arch/powerpc/platforms/pseries/Makefile
+++ b/arch/powerpc/platforms/pseries/Makefile
@@ -3,7 +3,7 @@ ccflags-$(CONFIG_PPC_PSERIES_DEBUG) += -DDEBUG
 
 obj-y   := lpar.o hvCall.o nvram.o reconfig.o \
       setup.o iommu.o event_sources.o ras.o \
-      firmware.o power.o dlpar.o mobility.o
+      firmware.o power.o dlpar.o mobility.o rng.o
 obj-$(CONFIG_SMP) += smp.o
 obj-$(CONFIG_SCANLOG) += scanlog.o
 obj-$(CONFIG_EEH) += eeh_pseries.o
diff --git a/arch/powerpc/platforms/pseries/rng.c b/arch/powerpc/platforms/pseries/rng.c
new file mode 100644
index 0000000..a702f1c
--- /dev/null
+++ b/arch/powerpc/platforms/pseries/rng.c
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2013, Michael Ellerman, IBM Corporation.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version
+ * 2 of the License, or (at your option) any later version.
+ */
+
+#define pr_fmt(fmt) "pseries-rng: " fmt
+
+#include <linux/kernel.h>
+#include <linux/of.h>
+#include <asm/archrandom.h>
+#include <asm/machdep.h>
+
+
+static int pseries_get_random_long(unsigned long *v)
+{
+ unsigned long retbuf[PLPAR_HCALL_BUFSIZE];
+
+ if (plpar_hcall(H_RANDOM, retbuf) == H_SUCCESS) {
+  *v = retbuf[0];
+  return 1;
+ }
+
+ return 0;
+}
+
+static __init int rng_init(void)
+{
+ struct device_node *dn;
+
+ dn = of_find_compatible_node(NULL, NULL, "ibm,random");
+ if (!dn)
+  return -ENODEV;
+
+ pr_info("Registering arch random hook.\n");
+
+ ppc_md.get_random_long = pseries_get_random_long;
+
+ return 0;
+}
+subsys_initcall(rng_init);
-- 
1.7.1